File: internal/opml.go
package internal

import (
	"encoding/xml"
	"time"
)

type OPML struct {
	XMLName xml.Name `xml:"opml"`
	Version string   `xml:"version,attr"`
	Head    Head     `xml:"head"`
	Body    Body     `xml:"body"`
}

type Head struct {
	Title        string `xml:"title"`
	DateCreated  string `xml:"dateCreated"`
	DateModified string `xml:"dateModified"`
}

type Body struct {
	Outlines []Outline `xml:"outline"`
}

type Outline struct {
	Text     string    `xml:"text,attr"`
	Type     string    `xml:"type,attr"`
	XMLURL   string    `xml:"xmlUrl,attr"`
	HTMLURL  string    `xml:"htmlUrl,attr,omitempty"`
	Language string    `xml:"language,attr,omitempty"`
	Title    string    `xml:"title,attr,omitempty"`
	Outlines []Outline `xml:"outline"`
}

func parseOPML(data []byte) (*OPML, error) {
	var opml OPML
	if err := xml.Unmarshal(data, &opml); err != nil {
		return nil, err
	}

	return &opml, nil
}

func exportOPML(feeds []Feed) ([]byte, error) {
	outlines := make([]Outline, len(feeds))
	for i, feed := range feeds {
		outlines[i] = Outline{
			Title:  feed.Title,
			Text:   feed.Title,
			Type:   "rss",
			XMLURL: feed.URL,
		}
	}

	opml := OPML{
		Version: "2.0",
		Head: Head{
			Title:        "My Feeds",
			DateCreated:  time.Now().Format(time.RFC1123Z),
			DateModified: time.Now().Format(time.RFC1123Z),
		},
		Body: Body{
			Outlines: outlines,
		},
	}

	bytes, err := xml.MarshalIndent(opml, "", "  ")
	if err != nil {
		return nil, err
	}

	return bytes, nil
}

File: src/lib/opml.ts
interface OPMLFeed {
  title: string
  xmlUrl: string
}

export function parseOPML(xml: string): OPMLFeed[] {
  const feeds: OPMLFeed[] = []
  const outlineRegex = /<outline[^>]*>/gi
  const matches = xml.match(outlineRegex) || []

  for (const match of matches) {
    const xmlUrlMatch = match.match(/xmlUrl=["']([^"']+)["']/i)
    const titleMatch = match.match(/(?:title|text)=["']([^"']+)["']/i)

    if (xmlUrlMatch) {
      feeds.push({
        title: titleMatch?.[1] || xmlUrlMatch[1],
        xmlUrl: xmlUrlMatch[1],
      })
    }
  }

  return feeds
}

export function generateOPML(feeds: { title: string; url: string }[]): string {
  const outlines = feeds
    .map(
      (f) =>
        `    <outline type="rss" text="${escapeXml(f.title)}" title="${escapeXml(f.title)}" xmlUrl="${escapeXml(f.url)}" />`
    )
    .join("\n")

  return `<?xml version="1.0" encoding="UTF-8"?>
<opml version="2.0">
  <head>
    <title>RSSy Subscriptions</title>
    <dateCreated>${new Date().toISOString()}</dateCreated>
  </head>
  <body>
${outlines}
  </body>
</opml>`
}

function escapeXml(str: string): string {
  return str
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;")
}
